index.bs

```diff
@@ -12839,6 +12871,34 @@ running the algorithm for an {{AudioNode}}, using an <a>input
 buffer</a> and the value(s) of the {{AudioParam}}(s) of this
 {{AudioNode}} as the input for this algorithm.
 
+
+<h3 id="error-handling-on-a-running-audio-context">
+    Error Handling for System Audio Resources on a running AudioContext</h3>
+
+In the event of an audio system resource error, the <a>rendering thread</a>
+performs the following steps upon receiving the error:
+
+1. Set {{[[rendering thread state]]}} of the {{AudioContext}} to
+    "{{AudioContextState/suspended}}".
+
+1. <a href="https://html.spec.whatwg.org/multipage/media.html#queue-a-media-element-task">
+    Queue a media element task</a> to execute the following steps:
+
+    1. <a spec="dom" lt="fire an event">fire an event</a> named
+        {{AudioContext/error}} at the associated {{AudioContext}}.
+    
+    1. If the {{BaseAudioContext/state}} attribute of the {{AudioContext}} is
+        not already "{{AudioContextState/suspended}}":
+
+        1. Set the {{BaseAudioContext/state}} attribute of the {{AudioContext}}
+            to "{{AudioContextState/suspended}}".
+
+        1. <a spec="dom" lt="fire an event">Fire an event</a> named
+            {{BaseAudioContext/statechange}} at the associated {{AudioContext}}.
+
+Note: An example of such a case would be when an external or wireless audio
+    device becomes disconnected during the active rendering of the AudioContext.
```

Comment:
Do we want to explicitly mention the fallbacks that we've all implemented here?

Reply:
Can you elaborate a bit more? Do you mean the fallback to the default device when no sinkID is given?

Reply:
I meant that when no devices are present on the system, as things stand today, Firefox (and I think Chrome, but I could have misheard) will run the graph on a regular thread, periodically attempting to grab an audio device should one appear.

Reply:
SG. I'll add that to the note.

Reply:
I gave more thoughts on this and the first part makes sense (`will run the graph on a regular thread`). However, the second part seems difficult to spec and expect consistent behavior between different implementations.

How does it "attempt" and how "periodic" should it be?

Reply:
By the way, I took a stab on this at L2012 above.

Reply:
I have no idea, but I can tell you what Firefox does.

If it cannot open a system level audio stream, it's going to attempt in 10ms. If that fails, it's going to backoff a bit and retry. The backoff progression is exponential but is capped at trying once a second in the end. While the retries are happening, the web audio rendering graph runs with a quantum of 10ms at the graph sample-rate, on a regular thread (not an audio IO thread).

Reply:
I see. 10ms is rather quite fast for retry! In Chromium, we simply spawn up a thread to emulate the device callback. The automatic retry seems like a a nice service, but I don't think we need to include in our spec language.

IMO, as a common ground what I have above makes sense: "the User-Agent must attempt to silently render the audio graph using a mechanism that emulates an audio output device."

WDYT?

Reply:
Yeah, we've found that trying very often and then backing off ended up with a good UX. Your phrasing sounds good to me.

Reply:
Great. Resolving.